### PR TITLE
Allow invisible anchors to be edited and deleted

### PR DESCRIPTION
**Pull request by mstrelan**
Uses an inline widget as per https://ckeditor.com/docs/ckeditor5/latest/tutorials/widgets/implementing-an-inline-widget.html to allow the anchor to be edited and deleted.

Might be possible to fix in the upstream repo.

https://github.com/northernco/ckeditor5-anchor-drupal/assets/665110/10de54ab-43c7-4dca-9257-e4d4755dadc7

### CODE REVIEW

File: theme/anchor.css [hunk 1]
```diff
@@ -8,3 +8,8 @@
  * it acts as a message to the builder telling that it should look for the corresponding styles
  * **in the theme** when compiling the editor.
  */
+.ck-anchor-placeholder {
+    color: dimgrey;
```
Comment (vtulse) [inline theme/anchor.css]: Please remove the formating changes of the anchor keep it as orignal plugin

File: src/utils.js [hunk 2]
```diff
@@ -72,18 +73,15 @@ export function createEmptyAnchorElement( id, { writer } ) {
...
-       domDocument.innerHTML = `[INVISIBLE ANCHOR: ${id}]`;
-    });
-    writer.addClass("ck-anchor", anchorElement);
-	writer.setCustomProperty( 'anchor', true, anchorElement );
-
-	return anchorElement;
+export function createEmptyPlaceholderAnchorElement( anchorId, { writer } ) {
+	const anchorElement = writer.createContainerElement('span', {
+		class: 'ck-anchor-placeholder',
+	}, [writer.createText(`#${anchorId}`)]);
```
Comment (vtulse) [inline src/utils.js]: Kepp the Invisible word
```
const anchorElement = writer.createContainerElement('span', {
     class: 'ck-anchor-placeholder',
}, [writer.createText(`[INVISIBLE ANCHOR: ${anchorId}]`)]);
```

File: src/anchorcommand.js [hunk 3]
```diff
@@ -207,6 +207,12 @@ export default class AnchorCommand extends Command {
 				[ 'anchorId', ...truthyManualDecorators, ...falsyManualDecorators ].forEach( item => {
 					writer.removeSelectionAttribute( item );
 				} );
+			} else if (selection.getSelectedElement()?.name === 'anchor') {
+				// Replace an invisible anchor.
```
Comment (vtulse) [inline src/anchorcommand.js:211]: Need to add the logic to remove the anchor if user has saved the empty ID while updating
```
// Replace an invisible anchor.
        if(id) {
          const anchor = writer.createElement('anchor', {
            anchorId: id
          });
          model.insertObject(anchor, null, null, {setSelection: 'on'});
        }
        else {
          // Remove an invisible anchor.
          if (selection.getSelectedElement()?.name === 'anchor') {
            model.deleteContent(selection);
          }
        }
```
Else we need to add empty input box validation with trim()

**Comment by vtulse** [comment]
Please add the red flag as icon insated of link icon which may confuse editor while editing Anchor Id vs Normal anchor
```
div[contenteditable=true] .ck-anchor {
	background: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAABmJLR0QA/wD/AP+gvaeTAAAAB3RJTUUH4AsCDSEAgw7OCAAAA1lJREFUWMPdl82LHEUYxn/VPTthe5Q1SMgt7GlZRG8KXrwEsgQlmEvwkIsL2YPgvyAEIoFcAvG0Nw8hgroQiHtIRNBLIAhqAlmQfHkIHiQssjg7n131eOjqmeqPnWF24x58oJjp6n7f99dVXdVPI4mpzTmctbg0LTRZi5ybHl9snykTkiicdM5hrcVai8sTW4vt95H0tqTvJP0pyaqotRFMHdC4/9MgpgJQq3+2tvY8F+impCNuMCiOjHPIWtxwiKSPSjFjAH+3k+WcHqyu6vujR/XD8eP65exZ7dy7V77qhaT3bLeLGw5xwyFpu42kz2vyjQGstSMAQFDkeXzxom6AvgZt+PYN6Ib//+zKlTrk3yX9Fnb8uLg4rp/VxOQAcRwLIEkSBHQ7HQDunzvH1sYGCdAAIjIJcIAFBr7vjdVV3lpfh2aTUP3nz/lpeZlnnQ6fSFm8tcbEcRVgYWEBgJ2dHQC+NIbXgLmgeKgQJPUw88DrS0tEzSbbDx/SAWKgA3ycA6SpMXFMo5zQGIOC44YPrisOYILzDeCIB9p+9Ah8/yu+r18T36hLKueYVca3vGjdSNWpAuCcI4rGKQyza5aYCmwURThphhQHUwVAh1i8FuCwVQFw+3gA/18jEEXRvpbhSwM47Cmo7AOSMME+oOC3vD7CzeelAdTJMd7vre+LyLbgfBueBKPS71SAcBpSoEu2xy+vrXHszBnM3Bx/373LH1evst3p0GT8ssphwuLyedLC8BkwpgoQRRE2AOgCb66s8M6dO4Xrjp0+zdKlS7h2m1/Pn+fJrVtEHiQOIJwv3ANOXrtWBIBsztM0HTmQVqulpNUaGYefT50KPcV9SR/4tl52IE8uX9ZN0HXQV75dB22Cdp8+DS9dKTiiCkCS1DmcD22/P3bFwyG220XSu96KjWR7Pb24fVt/bW7K7u6W85xwacqsACdHQWHzVt0OBkh6VdK3E1zlF2m7TVi8FiBJEs3Pz4eBF/a024Hldt792l4vn9r3/VQt2l5vz++IiiVrtVoYY2i32/mKMKE/mKjsjrLllr9VjcHkD5ypLtR6RxS8kmd6PfuldSBDkn8RHZYqI5Cm6dQgEwzlfmHzHI1ywjKAKc3btON9jcCkOzpogWmKphX5rwH+BWxcUbVSlumsAAAAJXRFWHRkYXRlOmNyZWF0ZQAyMDE2LTExLTAyVDEzOjMzOjAwKzAxOjAw1wGs7QAAACV0RVh0ZGF0ZTptb2RpZnkAMjAxNi0xMS0wMlQxMzozMzowMCswMTowMKZcFFEAAAAASUVORK5CYII=") no-repeat left center;
	background-size: 16px;
	padding-left: 18px;
	color: #00f;
	cursor: auto;
}
```

File: src/utils.js (hunk 2 above)
Comment (mstrelan) [reply src/utils.js]: I think we should use CSS to add `[INVISIBLE ANCHOR: ` prefix and `]` suffix so people can override it easily. Perhaps I'll raise a new issue for that after reverting this.

**Comment by mstrelan** [comment]
> Please add the red flag as icon insated of link icon which may confuse editor while editing Anchor Id vs Normal anchor

I agree with this but think it should be moved to a separate issue. See https://github.com/northernco/ckeditor5-anchor-drupal/pull/6

File: src/anchorcommand.js (hunk 3 above)
Comment (mstrelan) [reply src/anchorcommand.js:211]: Actually this is the same behaviour as visible anchors. I think this should be moved to a separate issue, and I think we can simply add the `required` attribute on the input element.

**Comment by ddonkersgoed-northern** [comment]
Thanks @mstrelan, this looks and works great. This was probably the most annoying issue left with the plugin and I'm glad to see it fixed. Merging this now.

**Pull request merged**